### ISSUE TITLE
[TypeScript] Add __typename to typings

### ISSUE THREAD

**Issue by eljenso** [dotansimha/graphql-code-generator#649]
I use GraphQL interfaces.
When querying for a field that might be any type implementing a specific interface I need to know which implementation it is, for this I currently use `__typename`.

Would be great if you would add the `__typename` field for all typings (should probably be an optional field).

**Comment by dotansimha** [comment]
@eljenso can you please provide an example for where it's missing? 
We added `__typename` for all client-side generated interfaces...

**Comment by eljenso** [comment]
Here is part of my schema: https://gist.github.com/eljenso/b2b43f30cdc3e9aa765f10df99de67a3
Running `apollo schema:download --endpoint 'http://localhost:4000'` yields this `schema.json`: https://gist.github.com/eljenso/6a8a53ce5b36edb19fb3d7a6fd3913cb
And after running `npx gql-gen --schema schema.json --template graphql-codegen-typescript-template --out ./src/typings/`, the resulting `types.ts` is this: https://gist.github.com/eljenso/5139abfbf4bed47c705120f5de46bedc

As you can see, interfaces `IElement` and `Section` do not have a `__typename` field.

**Comment by dotansimha** [comment]
@eljenso Can you explain you usage of `__typename` in the generated typings for the server side? They are generated automatically by `graphql` and you shouldn't deal with them as far as I know.

The `__typename` is an internal field, and usually it's usage is for client-side to know what is the declared type of each structure. 

Btw, you don't have to download the schema first, you can just use `--schema http://localhost:4000` ;)

**Comment by eljenso** [comment]
> The __typename is an internal field, and usually it's usage is for client-side to know what is the declared type of each structure.

This is exactly what I want to do but the typings generated by `graphql-codegen-typescript-template` do not have `__typename`.
I have some GraphQL types which do only differ in name (i.e. fields are all the same), so I need `__typename` client-side in order to differentiate between those TS interfaces.

**Comment by jonaskello** [comment]
I would also like `__typename` to be optionally added to the full schema types. My use-case is that I have my own client-side caching solution and the ability to modify the cache (for optimistic updates). So when I modify the cache on the client-side I use the full schema types. This is done because the optimistic update is allowed to add any field within the full type, but no fields that are not in the full type. So for this reason I'm using the full types on the client-side. Not for queries but strictly for cache manipulation.

**Comment by Elijen** [comment]
`__typename` has many usages in client applications. It's also a valid field in a GraphQL query. I think it would be really nice to have the option to add it into all types (even though I think it should be done by default).

@eljenso Any chance you could re-open the issue?

**Comment by jonaskello** [comment]
@Elijen see also #1633, and [this comment](https://github.com/dotansimha/graphql-code-generator/issues/1633#issuecomment-479388007) in that issue.

**Comment by Elijen** [comment]
@jonaskello Thanks, I've deleted my comment because I though I've misunderstood the behavior, but it seems generating `__typename` would really be helpful.

@dotansimha In your example I would expect the types look like this:
```
export type Comment = {
  __typename: "Comment",
  /** The SQL ID of this entry */
  id: Scalars['Int'],
  /** The GitHub user who posted the comment */
  postedBy: User,
  /** A timestamp of when the comment was posted */
  createdAt: Scalars['Float'],
  /** The text of the comment */
  content: Scalars['String'],
  /** The repository which this comment is about */
  repoName: Scalars['String'],
};
```

I think it makes more sense as this is how the object actually looks like.

**Comment by leebenson** [comment]
Also https://github.com/dotansimha/graphql-code-generator/issues/1633#issuecomment-479850723

We could very much do with `__typename` being selectable in a returned response.

**Comment by Elijen** [comment]
> Also [#1633 (comment)](https://github.com/dotansimha/graphql-code-generator/issues/1633#issuecomment-479850723)
> 
> We could very much do with `__typename` being selectable in a returned response.

Keep in mind that for example `apollo-client` always selects `__typename` even if it's not specified in the query.

**Comment by leebenson** [comment]
To provide a use-case, re:

> @eljenso Can you explain you usage of __typename in the generated typings for the server side? They are generated automatically by graphql and you shouldn't deal with them as far as I know.

☝️ Despite being an internal field, it's still very useful to have access to this in user-land code.

We have a ton of interfaces, and require knowing which concrete type is being returned.

We also pass around `__typename` in child React components to determine which subsequent GraphQL queries should be executed based on that concrete type.

IMO, it should be a selectable field like any other.

**Comment by leebenson** [comment]
> Keep in mind that for example apollo-client always selects __typename even if it's not specified in the query.

This is true, but because GraphQL Code Generator doesn't include `__typename` on individual fields, TS only reports it being available at the top-level.

Example --

Given something like this:

```gql
fragment Source on Source {
  __typename  # <-- explicitly requested on the fragment
  id
  name
  type
}

query Builder(
  $source: Boolean!
  $sourceId: ID!
) {

  # ... other types omitted for brevity...

  source: node(id: $sourceId) @include(if: $source) {
    __typename # <-- explicitly requested at the root-level, e.g. same as the fragment
    ...BuilderSource
  }
}

```

I get `__typename` on the root-level query:

<img width="472" alt="Screenshot 2019-04-04 at 12 14 01" src="https://user-images.githubusercontent.com/719722/55551649-3d6aac00-56d3-11e9-84c6-e40ed62ac950.png">

But not on the individual `source` field, 

<img width="410" alt="Screenshot 2019-04-04 at 12 15 21" src="https://user-images.githubusercontent.com/719722/55551698-570bf380-56d3-11e9-9b83-ae59541bfbf1.png">

**TL;DR - We should get `__typename` on fields, too.**

**Comment by dotansimha** [comment]
@leebenson I think in the most recent version (>1.0), it should be there. All levels of selection set should have `__typename`. If it's specified in the selection set, it will be required, otherwise, it will be optional. Is that's no the behaviour you are aiming for, can you please open a new issue? 

Regarding adding `__typename` to the base types, there is an open issue for that, and we'll try to address it soon.

**Comment by leebenson** [comment]
@dotansimha the screenshots in https://github.com/dotansimha/graphql-code-generator/issues/649#issuecomment-479855997 are actually generated from v1.0.2, so it looks like this is still an issue.

It's definitely specified in the selection set - both in the `fragment` as well as an explicit field (just in case that first one in the fragment wasn't being picked up.)

**Comment by dotansimha** [comment]
@leebenson I tried to reproduce it in our live-demo.

Test schema:
```graphql
interface Node {
  id: ID!
}

type Source implements Node {
  id: ID!
  name: String!
  type: String!
}

type Query {
  node(id: ID!): Node!
}
```

Test query:
```graphql
fragment Source on Source {
  __typename  # <-- explicitly requested on the fragment
  id
  name
  type
}

query Builder(
  $source: Boolean!
  $sourceId: ID!
) {

  # ... other types omitted for brevity...

  source: node(id: $sourceId) @include(if: $source) {
    __typename # <-- explicitly requested at the root-level, e.g. same as the fragment
    ...Source
  }
}
```

And using `typescript` and `typescript-operations`, I'm getting this output:

![image](https://user-images.githubusercontent.com/3680083/55684690-dc530a80-5956-11e9-8af9-197d465bf3ba.png)

```typescript
type Maybe<T> = T | null;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string,
  String: string,
  Boolean: boolean,
  Int: number,
  Float: number,
};

export type Node = {
  id: Scalars['ID'],
};

export type Query = {
  node: Node,
};


export type QueryNodeArgs = {
  id: Scalars['ID']
};

export type Source = Node & {
  id: Scalars['ID'],
  name: Scalars['String'],
  type: Scalars['String'],
};
export type SourceFragment = ({ __typename: 'Source' } & Pick<Source, 'id' | 'name' | 'type'>);

export type BuilderQueryVariables = {
  source: Scalars['Boolean'],
  sourceId: Scalars['ID']
};


export type BuilderQuery = ({ __typename?: 'Query' } & { source: SourceFragment });
```

As you can see, the `SourceFragment` type has `__typename` field set. 

I tried to copy the content to TypeScript Playground, create a fake object based on the response, and then try to access it: 

![image](https://user-images.githubusercontent.com/3680083/55684762-c7c34200-5957-11e9-8612-0baf846bacfc.png)

Everything seems to work. I also tried to remove `__typename` explicitly from the `query`, and now the `__typename` properties are being added, but as optional:

```typescript
export type SourceFragment = ({ __typename?: 'Source' } & Pick<Source, 'id' | 'name' | 'type'>);
export type BuilderQuery = ({ __typename?: 'Query' } & { source: SourceFragment });
```

And it's still works:

![image](https://user-images.githubusercontent.com/3680083/55684802-183a9f80-5958-11e9-8292-f92028d03715.png)

Maybe I'm missing something, or something is different in the schema or something like that, but I think it should work just fine. If possible, @leebenson , can you please share the actual generated types you are getting from the codegen?

**Comment by leebenson** [comment]
@dotansimha - thanks for your example, and apologies for the delay in getting back to you.

The error seems to be with interface types:

**Schema:**

```gql
interface Node {
  id: ID!
}

type A implements Node {
  id: ID!
  A: String
}

type B implements Node {
  id: ID!
  B: String
}

type Query {
  some: Node
}
```

**Documents:**

```gql
fragment Node on Node {
  __typename
  id
}

query Test {
  some {
    ...Node
  }
}
```

**Config:**

```yml
generates:
  server-types.ts:
    - typescript
    - typescript-operations
```

---

Generated:

<img width="374" alt="Screenshot 2019-04-28 at 20 51 30" src="https://user-images.githubusercontent.com/719722/56869338-70187380-69f7-11e9-8b58-05c5c3d47b4a.png">

Note the lack of `__typename`.

In this case, I'd expect it to be closer to:

```ts
export type NodeFragment = ({ __typename: 'A' | 'B'} & Pick<Node, "id">)
```

**Comment by dotansimha** [comment]
You are right, @leebenson , I'm working on a fix. Thank you!
